### PR TITLE
HaProxy 2.4-dev18 support

### PR DESCRIPTION
**Pull request by julek-wolfssl**
*This patch is dependent on https://github.com/wolfSSL/wolfssl/pull/3871 because proto version selection logic is refactored in that pull request.*
This patch contains the following changes:
- Enable more options with `--enable-haproxy`
- Compatibility layer additions
    - `STACK_TYPE_X509_OBJ`
    - `OCSP_id_cmp`
    - `X509_STORE_get0_objects`
    - `X509V3_EXT_nconf_nid`
    - `X509V3_EXT_nconf`
    - `X509_chain_up_ref`
    - `X509_NAME_hash`
    - `sk_X509_NAME_new_null`
    - `X509_OBJECT_get0_X509`
    - `X509_OBJECT_get0_X509_CRL`
    - `ASN1_OCTET_STRING_free`
    - `X509_LOOKUP_TYPE`
    - `OSSL_HANDSHAKE_STATE`
- New `OPENSSL_COMPATIBLE_DEFAULTS` define will set default behaviour that is compatible with OpenSSL
    - WOLFSSL_CTX
        - Enable all compiled in protocols
        - Allow anonymous ciphers
        - Set message grouping
        - Set verify to SSL_VERIFY_NONE
- In `SetSSL_CTX`, don't change `send` and `recv` callback if currently using `BIO`
- `ssl->peerVerifyRet`
    - Return first that occured
    - Set correct value on date error
    - Set revoked error on OCSP or CRL error
    - Save value in session and restore on resumption
    - Add to session serialization
- With `OPENSSL_EXTRA`, send an alert on invalid downgrade attempt
- Handle sni callback `SSL_TLSEXT_ERR_NOACK`
- Add `WOLFSSL_VERIFY_DEFAULT` option for `wolfSSL_CTX_set_verify` and `wolfSSL_set_verify` to allow resetting to default behaviour

### CODE REVIEW

**Comment by julek-wolfssl** [comment]
Retest this please.

File: wolfssl/ssl.h [hunk 1]
```diff
@@ -2159,6 +2180,7 @@ enum { /* ssl Constants */
     WOLFSSL_FILETYPE_DEFAULT = 2, /* ASN1 */
     WOLFSSL_FILETYPE_RAW     = 3, /* NTRU raw key blob */
 
+    WOLFSSL_VERIFY_DEFAULT              = -1,
```
Comment (SparkiDev) [inline wolfssl/ssl.h:2183]: WOLFSSL_VERIFY_* are bits
WOLFSSL_VERIFY_DEFAULT should have the bits set that reflect what is to be verified by default.

It appears you are actually trying to avoid setting the verifyNone field on the WOLFSSL context. Is that right?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ocsp.c [hunk 2]
```diff
@@ -1021,6 +1021,40 @@ const WOLFSSL_OCSP_CERTID* wolfSSL_OCSP_SINGLERESP_get0_id(const WOLFSSL_OCSP_SI
...
+    int ret = 0;
+    if (a == NULL || b == NULL)
+        return WOLFSSL_FATAL_ERROR;
+
+    ret = a->hashAlgoOID != b->hashAlgoOID;
+    if (ret == 0)
+        ret = XMEMCMP(a->issuerHash, b->issuerHash, OCSP_DIGEST_SIZE);
+    if (ret == 0)
+        ret = XMEMCMP(a->issuerKeyHash, b->issuerKeyHash, OCSP_DIGEST_SIZE);
+    if (ret == 0) {
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ocsp.c:1042]: Won't the first if statement set an error when `a->status` and `b->status` point to different objects? I assume that this is a valid case. If they would have to point to the same object then checking the serial would be unnecessary.

File: src/ssl.c [hunk 3]
```diff
@@ -429,6 +436,23 @@ WOLFSSL_CTX* wolfSSL_CTX_new_ex(WOLFSSL_METHOD* method, void* heap)
...
 
+#ifdef OPENSSL_COMPATIBLE_DEFAULTS
+    if (ctx) {
+        if (wolfSSL_CTX_set_min_proto_version(ctx,
+                SSL3_VERSION) != WOLFSSL_SUCCESS ||
+#ifdef HAVE_ANON
+                wolfSSL_CTX_allow_anon_cipher(ctx) != WOLFSSL_SUCCESS ||
+#endif
+                wolfSSL_CTX_set_group_messages(ctx) != WOLFSSL_SUCCESS
+                ) {
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:447]: I'm not sure what you mean by this comment.

File: wolfssl/ssl.h (hunk 1 above)
Comment (julek-wolfssl) [reply wolfssl/ssl.h]: I am trying to restore it back to the defaults set before `wolfSSL_CTX_set_verify` is called. I changed how the `WOLFSSL_VERIFY_*` bits are defined and changed `WOLFSSL_VERIFY_DEFAULT` to a higher bit value.

File: src/internal.c [hunk 4]
```diff
@@ -11388,7 +11394,8 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
 
             #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
                 if (args->totalCerts >= MAX_CHAIN_DEPTH) {
-                    ssl->peerVerifyRet = X509_V_ERR_CERT_CHAIN_TOO_LONG;
+                    if (!ssl->peerVerifyRet) /* Return first cert error here */
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/internal.c:11397]: Changed.

File: src/internal.c [hunk 5]
```diff
@@ -11840,15 +11850,24 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
...
-                    ssl->peerVerifyRet = X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE;
-                #endif
+                    #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+                    if (!ssl->peerVerifyRet) { /* Return first cert error here */
+                        if (ret == ASN_BEFORE_DATE_E)
+                            ssl->peerVerifyRet = X509_V_ERR_CERT_NOT_YET_VALID;
+                        else if (ret == ASN_AFTER_DATE_E)
+                            ssl->peerVerifyRet = X509_V_ERR_CERT_HAS_EXPIRED;
+                        else
+                            ssl->peerVerifyRet =
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/internal.c:11867]: Changed.

File: wolfssl/wolfcrypt/asn.h [hunk 6]
```diff
@@ -316,7 +316,11 @@ enum Misc_ASN {
     MAX_DATE_SIZE       =  32,
     ASN_GEN_TIME_SZ     =  15,     /* 7 numbers * 2 + Zulu tag */
 #ifndef NO_RSA
+#ifdef WOLFSSL_HAPROXY
+    MAX_ENCODED_SIG_SZ  = 1024,
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfssl/wolfcrypt/asn.h]: Comments added.

File: wolfssl/ssl.h [hunk 7]
```diff
@@ -2001,28 +2021,29 @@ enum {
...
+    X509_V_ERR_CERT_REVOKED                      = 23,
+    X509_V_ERR_CERT_REJECTED,
     /* Required for Nginx  */
-    X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT       = 24,
-    X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN         = 25,
-    X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY = 26,
-    X509_V_ERR_CERT_UNTRUSTED                    = 27,
-    X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE   = 28,
-    X509_V_ERR_SUBJECT_ISSUER_MISMATCH           = 29,
+    X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT,
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfssl/ssl.h]: Put number back.

File: wolfssl/ssl.h [hunk 8]
```diff
@@ -1905,7 +1924,7 @@ enum {
     SSL_OP_DONT_INSERT_EMPTY_FRAGMENTS            = 0x00100000,
     SSL_OP_NO_QUERY_MTU                           = 0x00200000,
     SSL_OP_COOKIE_EXCHANGE                        = 0x00400000,
-    SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = 0x00800000,
+    WOLFSSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = 0x00800000,
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply wolfssl/ssl.h]: Haproxy was checking this in a pre-proc macro. I modified the Haproxy patch to omit this check for wolfSSL and changed this enum back to how it was.

File: src/ssl.c [hunk 9]
```diff
@@ -46246,6 +46492,42 @@ void wolfSSL_THREADID_set_numeric(void* id, unsigned long val)
 }
 #endif
 
+#ifdef OPENSSL_ALL
+WOLFSSL_X509_LOOKUP_TYPE wolfSSL_X509_OBJECT_get_type(
+        const WOLFSSL_X509_OBJECT* obj)
+{
+    if (obj != NULL)
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:46499]: Changed.

File: src/ssl.c [hunk 10]
```diff
@@ -46256,16 +46538,23 @@ WOLFSSL_X509_OBJECT* wolfSSL_sk_X509_OBJECT_delete(
...
 
-void wolfSSL_X509_OBJECT_free(WOLFSSL_X509_OBJECT *a)
+WOLFSSL_X509 *wolfSSL_X509_OBJECT_get0_X509(const WOLFSSL_X509_OBJECT *obj)
 {
-    WOLFSSL_ENTER("wolfSSL_X509_OBJECT_free");
-    WOLFSSL_STUB("wolfSSL_X509_OBJECT_free");
-    (void)a;
+    if (obj != NULL && obj->type == WOLFSSL_X509_LU_X509)
+        return obj->data.x509;
+    else
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:46547]: Changed.

File: src/ocsp.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/ocsp.c:1042]: Sorry. My change won't work.

File: src/ssl.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (SparkiDev) [reply src/ssl.c:447]: Move ') {' up to end of previous line if it fits with 80 characters.

File: wolfssl/ssl.h (hunk 1 above)
Comment (SparkiDev) [reply wolfssl/ssl.h]: Would this be better :
    ssl->options.verifyNone = (mode == WOLFSSL_VERIFY_NONE);
    ssl->options.verifyPeer = ((mode & WOLFSSL_VERIFY_PEER) == WOLFSSL_VERIFY_PEER);
...

File: src/ssl.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:447]: Moved.

File: wolfssl/ssl.h (hunk 1 above)
Comment (julek-wolfssl) [reply wolfssl/ssl.h]: Changed to your suggested method.

**Review by SparkiDev — APPROVED** [review]

File: wolfssl/openssl/ssl.h [hunk 11]
```diff
@@ -440,19 +444,17 @@ typedef STACK_OF(ACCESS_DESCRIPTION) AUTHORITY_INFO_ACCESS;
 #define X509_REQ_get_pubkey             wolfSSL_X509_get_pubkey
 #define X509_get_notBefore              wolfSSL_X509_get_notBefore
 #define X509_get0_notBefore             wolfSSL_X509_get_notBefore
-#define X509_getm_notBefore             wolfSSL_X509_get_notBefore
+#define X509_getm_notBefore             (ASN1_TIME*)wolfSSL_X509_get_notBefore
```
Comment (SparkiDev) [inline wolfssl/openssl/ssl.h:447]: wolfSSL_X509_get_notBefore() returns WOLFSSL_ASN1_TIME*.
ASN1_TIME is typedefed to WOLFSSL_AN1_TIME.
Why is this necessary?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/tls13.c [hunk 12]
```diff
@@ -4630,11 +4630,19 @@ int DoTls13ClientHello(WOLFSSL* ssl, const byte* input, word32* inOutIdx,
         if (!ssl->options.downgrade) {
             WOLFSSL_MSG("Client trying to connect with lesser version than "
                         "TLS v1.3");
+#ifdef OPENSSL_EXTRA
```
Comment (JacobBarthelmeh) [inline src/tls13.c:4633]: would this be something that could also fall under the macro WOLFSSL_EXTRA_ALERTS?

File: wolfssl/ssl.h [hunk 13]
```diff
@@ -1367,10 +1383,12 @@ WOLFSSL_API WOLFSSL_SESSION* wolfSSL_get1_session(WOLFSSL* ssl);
                            /* what's ref count */
 
 WOLFSSL_API WOLFSSL_X509* wolfSSL_X509_new(void);
-#if defined(OPENSSL_EXTRA_X509_SMALL) || defined(OPENSSL_ALL)
+#if defined(OPENSSL_EXTRA_X509_SMALL) || defined(OPENSSL_EXTRA)
```
Comment (JacobBarthelmeh) [inline wolfssl/ssl.h:1386]: why change from openssl_all to openssl_extra here? The define between the two is already grey ..but having the compiled functions switch periodically could make things not work for builds that used to use the functions.

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]

File: src/ssl.c [hunk 14]
```diff
@@ -9934,9 +9999,8 @@ void* wolfSSL_X509_get_ext_d2i(const WOLFSSL_X509* x509, int nid, int* c,
                     }
 
                     gn->type = dns->type;
-                    gn->d.ia5->length = dns->len;
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:9937]: `wolfSSL_ASN1_STRING_set` sets this. It is called immediately after so this value gets overwritten.

File: src/ssl.c [hunk 15]
```diff
@@ -30078,7 +30274,7 @@ int wolfSSL_sk_SSL_COMP_num(WOLF_STACK_OF(WOLFSSL_COMP)* sk)
  * the session to resume it.
  */
 void wolfSSL_CTX_sess_set_get_cb(WOLFSSL_CTX* ctx,
-                    WOLFSSL_SESSION*(*f)(WOLFSSL*, unsigned char*, int, int*))
+                    WOLFSSL_SESSION*(*f)(WOLFSSL*, const unsigned char*, int, int*))
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:30277]: Version 1.1.1 of OpenSSL adds the const attribute. https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_sess_set_get_cb.html

File: wolfssl/ssl.h (hunk 13 above)
Comment (julek-wolfssl) [reply wolfssl/ssl.h:1386]: My understanding of the defines is this: OPENSSL_EXTRA_X509_SMALL < OPENSSL_EXTRA < OPENSSL_ALL. With each level being a superset of the previous. This define was including the functions for OPENSSL_EXTRA_X509_SMALL but not for OPENSSL_EXTRA which is both confusing and doesn't follow the rule of each level including the previous. I changed this to be more consistent on what is and isn't available.

File: wolfssl/openssl/ssl.h (hunk 11 above)
Comment (julek-wolfssl) [reply wolfssl/openssl/ssl.h:447]: I'm not sure why I changed this. I've reverted it back to how it was.

File: src/internal.c [hunk 16]
```diff
@@ -8401,17 +8418,6 @@ static int wolfSSLReceive(WOLFSSL* ssl, byte* buf, word32 sz)
     if (recvd < 0) {
         switch (recvd) {
             case WOLFSSL_CBIO_ERR_GENERAL:        /* general/unknown error */
-                #if defined(OPENSSL_ALL) || defined(WOLFSSL_APACHE_HTTPD)
-                #ifndef NO_BIO
-                    if (ssl->biord) {
-                        /* If retry and read flags are set, return WANT_READ */
-                        if ((ssl->biord->flags & WOLFSSL_BIO_FLAG_READ) &&
-                            (ssl->biord->flags & WOLFSSL_BIO_FLAG_RETRY)) {
-                            return WANT_READ;
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/internal.c:8410]: Added back in for Apache.

File: src/ssl.c [hunk 17]
```diff
@@ -7618,6 +7640,49 @@ int wolfSSL_CTX_check_private_key(const WOLFSSL_CTX* ctx)
 }
 #endif /* !NO_CHECK_PRIVATE_KEY */
 
+#ifdef OPENSSL_ALL
+WOLFSSL_EVP_PKEY* wolfSSL_CTX_get0_privatekey(const WOLFSSL_CTX* ctx)
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:7644]: Added a comment about who is responsible for the returned struct.

File: src/ssl.c [hunk 18]
```diff
@@ -10338,7 +10402,120 @@ void wolfSSL_X509V3_set_ctx_nodb(WOLFSSL_X509V3_CTX* ctx)
...
+            WOLFSSL_MSG("invalid or unsupported NID");
+            goto err_cleanup;
+    }
+    return ext;
+err_cleanup:
+    wolfSSL_X509_EXTENSION_free(ext);
+    return NULL;
+}
+
+WOLFSSL_X509_EXTENSION* wolfSSL_X509V3_EXT_nconf_nid(WOLFSSL_CONF* conf,
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c:10474]: I added some documentation. These functions are very basic and don't perform much checking either so be sure to add testing in wolfCLU if you do decide to use the.

File: src/ssl.c [hunk 19]
```diff
@@ -17151,6 +17327,24 @@ int wolfSSL_CTX_set_max_proto_version(WOLFSSL_CTX* ctx, int ver)
     return CheckSslMethodVersion(ctx->method->version.major, ctx->mask);
 }
 
+int wolfSSL_set_min_proto_version(WOLFSSL* ssl, int ver)
```
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply src/ssl.c]: Implemented them now that the referenced PR has been merged.

File: src/tls13.c (hunk 12 above)
Comment (julek-wolfssl) [reply src/tls13.c]: Added new alerts to `WOLFSSL_EXTRA_ALERTS`.

**Review by SparkiDev — APPROVED** [review]

**Comment by SparkiDev** [comment]
Please fix: ./configure CFLAGS="-pedantic" --enable-all

**Comment by julek-wolfssl** [comment]
`-pedantic` flag satisfied.

File: wolfssl/ssl.h (hunk 13 above)
Comment (JacobBarthelmeh) [reply wolfssl/ssl.h:1386]: My concern is that the OPENSSL_EXTRA_X509_SMALL build changes else where in the code and this change from OPENSSL_ALL to OPENSSL_EXTRA might not be related to HAPROXY, but have the potential for causing issues with other existing builds. Checking the OPENSSL_EXTRA_X509_SMALL build after this PR the code size has increased some, which maybe unavoidable. Yes, the OPENSSL_EXTRA_X509_SMALL build is a subset of OPENSSL_EXTRA but it was also fine tuned initially for a specific use case with not a lot of memory (including memory taken up by code compiled into a static library).

It would be better I think to avoid changing what functions are compiled with OPENSSL_ALL vs OPENSSL_EXTRA vs OPENSSL_EXTRA_X509_SMALL if possible with the HAPROXY port update and have a separate PR that addresses refactoring that code.

Comment (julek-wolfssl) [reply wolfssl/ssl.h:1386]: I don't think I'm changing too much what gets included with `OPENSSL_EXTRA_X509_SMALL`. Many of the changes I made were for consistency. For example the reference counters were being inconsistently compiled in. I don't think these changes will break other builds (it might even address some issues).

> separate PR that addresses refactoring that code.

I prefer addressing small changes immediately as they come up. I think that these inconsistencies can be addressed directly by this PR.

**Review by JacobBarthelmeh — APPROVED** [review]
